src/dlfcn.c

```diff
@@ -513,10 +513,18 @@ static const char *get_symbol_name( HMODULE module, IMAGE_IMPORT_DESCRIPTOR *iid
             IMAGE_IMPORT_BY_NAME *nameData;
             if (IMAGE_SNAP_BY_ORDINAL(thunkILT->u1.Ordinal))
               continue;
-            if ((void *)thunkIAT->u1.Function != addr)
-              continue;
-            nameData = (IMAGE_IMPORT_BY_NAME*)(thunkILT->u1.AddressOfData + base);
-            return (const char *)nameData->Name;
+            if ( (void *)thunkIAT->u1.Function == addr ) {
+                 nameData = (IMAGE_IMPORT_BY_NAME*)( thunkILT->u1.AddressOfData + base );
+                 *func_address = (void *)thunkIAT->u1.Function;
+                 return (const char *)nameData->Name;
+            }
+
+            /* if address is not greater than function + 10 */
+            if ( addr > (void *)thunkIAT->u1.Function && addr < (void *)thunkIAT->u1.Function + 10 ) {

```

Comment:
This looks like a hack with arbitrary chosen constant 10.

Working linear-time algorithm is simple: Just choose symbols name which function address is the biggest one from the all addresses, but which is less or equal to the request address from dladdr() argument.

```c
void *candidateAddr = NULL;
const char* candidateName = NULL;
...
for(i = 0; iid[i].Characteristics != 0 && iid[i].FirstThunk != 0; i++) {
    ...
    if( (void *)thunkIAT->u1.Function <= addr && candidateAddr < (void *)thunkIAT->u1.Function ) {
        candidateAddr = (void *)thunkIAT->u1.Function;
        nameData = (IMAGE_IMPORT_BY_NAME*)( thunkILT->u1.AddressOfData + base );
        candidateName = (const char *)nameData->Name;
    }
}
*addr = candidateAddr;
return candidateName;
```

Reply:
This assume to have the addresses in the iat in ascending order - is that given ?

Reply:
No, this is not requirement. If you do not break a loop and iterate to its end then you can find the maximum value even when addresses are not sorted.